File: worker/src/telemetry.ts
import { NodeSDK } from "@opentelemetry/sdk-node";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { HttpInstrumentation } from "@opentelemetry/instrumentation-http";
import { ExpressInstrumentation } from "@opentelemetry/instrumentation-express";
import { resourceFromAttributes } from "@opentelemetry/resources";

const endpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT;

export const sdk = new NodeSDK({
  resource: resourceFromAttributes({ "service.name": "sovereign-worker" }),
  traceExporter: endpoint ? new OTLPTraceExporter({ url: endpoint }) : undefined,
  instrumentations: [new HttpInstrumentation(), new ExpressInstrumentation()]
});

export async function startTelemetry(): Promise<void> {
  await sdk.start();
}

export async function shutdownTelemetry(): Promise<void> {
  await sdk.shutdown();
}

File: vault/app/telemetry.py
from __future__ import annotations

import os

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor


def configure_tracing() -> None:
    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
    provider = TracerProvider(resource=Resource.create({"service.name": "sovereign-vault"}))
    if endpoint:
        provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter(endpoint=endpoint)))
    trace.set_tracer_provider(provider)
